Additional Texture Formats/Extensions Support

TDCRanila opened the pull request:
Heya!

I was messing around with the tool, but noticed that textures other then '.png' were not supported. From what I can gather from the code, this is just because the texture file format when loading is hard-coded to '.png'. Changing the extension to '.tga', for instance, worked perfectly fine. So because of this, I have added support for other file extensions. 

I decided to let the user control the array of strings/extensions, so that they order, remove, or add the ones they needed for their map. (Not sure if this really needed though and I am not convinced that is the greatest UI/UX experience, however, it works fine.)
Alternatively, we could:
- Use booleans to represent the supported texture/image extensions - although it could clutter the code somewhat. 
- Just use the list of extensions that Godot supports via the "resource_loader->get_recognized_extensions_for_type" function instead and then remove the "Texture Import Extensions" property entirely. Although, it can happen that when the map is looking for a texture with an extension that is last on that extension list, it will waste processing time going through the others.
- ...or something else that I haven't thought of.

Anyhow, this is how you can use it in the editor.

https://user-images.githubusercontent.com/38665664/190424509-b89a3c4c-18bd-4736-8d01-6482ae7cf41d.mp4

I have also made a change regarding "texture_from_name" so that we don't have to; check for file extensions, check if it exists, and then call the load function for every texture. (This would be done twice for setting the texture size and then later for the mesh building.) It now fetches the resource from "m_loaded_map_textures" dictionary, which gets filled once at the start of map load.

Of course, feel free to suggest anything and let me know what you think. :)

TDCRanila commented:
Alrighty, I did some research, 

- Godot's Extensions List for "CompressedTexture2D" returns a list that contains the following supported extensions; 
```tres, res, png, bmp, hdr, jpg, jpeg, svg, tga, exr, webp, ctex, dds```

- Trenchbroom states in their [documentation](https://trenchbroom.github.io/manual/latest/#game_configuration_files): 

> "The _'image'_ format can be used to load a wide array of image formats such as tga, pcx, jpeg, and so on. **_TrenchBroom uses the FreeImage Library to load these images and supports any file type supported by this library._**"

[FreeImage Library Supported Formats](https://freeimage.sourceforge.io/features.html)
![freeimage-supported-formats](https://user-images.githubusercontent.com/38665664/190680587-dfec22ad-2ff8-4da4-ac6e-766e6cf29732.png)

So theoretically, that would mean that all supported texture formats/extensions overlap besides ```tres, res, and svg```, and thus means that whatever you are able to use in Trenchbroom, Godot will also be able to load it - using directory-based texture collections of course.

As for the PR, I would then indeed just use the Godot's supported extension list and then remove all the bits with the exposed property. 
If we want to be more explicit with the extension list, it could be hardcoded, but I don't see much value in that since that Godot function exist albeit with extensions that Trenchbroom can't load.

On a sidenote, if we want to support more texture formats/extensions in the workflow, the 'textures' field in GameConfig.cfg that is provided with TBLoader, would have to be updated as well to accommodate this.

codecat commented:
Thanks for taking the time to look into this!

I think I would have a preference for hardcoding the list to take into account the (unlikely?) scenario that Godot adds another format that is unsupported by Trenchbroom, or one that could be something *other* than an image (like `tres`). Then if someone needs another format we could just add it on request, since that would probably be a rare occasion as well.

And yes indeed, the default `GameConfig.cfg` would have to be updated as well. Since we don't generate this file yet, hardcoding the supported extensions also makes more sense to me.

TDCRanila commented:
Yeah, no problem! ^^

I have removed the 'Texture Import Extensions' property entirely and replaced it with a hardcoded list instead like we talked about.

Let me know if there is anything else that could be improved or changed.

codecat commented on src/builder.cpp:
```diff
@@ -509,10 +510,42 @@ MeshInstance3D* Builder::build_entity_mesh(int idx, LMEntity& ent, Node3D* paren
+	
+	// Attempt to load and cache textures used by the map
+	auto resource_loader = ResourceLoader::get_singleton();
+	const LMTextureData* tex;
+	String tex_path;
```
These 2 variables could be scoped inside of the for loop below, I think.

TDCRanila replied:
Yeah, that would work fine, but I thought since we are looping a lot for every texture and extension, it would be better to overwrite them instead of allocating/deallocating them constantly. Just a small thing really, but if you prefer otherwise, that's cool too.

codecat replied:
Agreed! Probably not super important for the pointer though, but for the string it does make sense.

codecat commented on src/builder.cpp:
```diff
@@ -509,10 +510,42 @@ MeshInstance3D* Builder::build_entity_mesh(int idx, LMEntity& ent, Node3D* paren
+	String tex_path;
+
+	for (int tex_i = 0; tex_i < m_map->texture_count; tex_i++) {
+		bool has_loaded_texture(false);
```
This might look nicer? 😊
```cpp
bool has_loaded_texture = false;
```

TDCRanila replied:
Oops! That slipped through, will correct that.

codecat commented on src/builder.cpp:
```diff
@@ -436,18 +438,17 @@ MeshInstance3D* Builder::build_entity_mesh(int idx, LMEntity& ent, Node3D* paren
-			auto res_texture = texture_from_name(name);
-
+			auto res_texture = texture_from_name(tex.name);
+			
```
Some stray whitespace here

codecat reviewed (changes requested):
I just have a couple little style nitpicks, otherwise it looks great! 👍

TDCRanila commented:
Alrighty, I have made the adjustments and have amended the commit, so that should be it. 🎉

codecat commented:
Looks good to me! Thank you! 🎉
